plugin/evm/config/config.md
In order to specify a config for the C-Chain, a JSON config file should be placed at `{chain-config-dir}/C/config.json`. This file does not exist by default.

For example if `chain-config-dir` has the default value which is `$HOME/.luxd/configs/chains`, then `config.json` should be placed at `$HOME/.luxd/configs/chains/C/config.json`.

The C-Chain config is printed out in the log when a node starts. Default values for each config flag are specified below.

Default values are overridden only if specified in the given config file. It is recommended to only provide values which are different from the default, as that makes the config more resilient to future default changes. Otherwise, if defaults change, your node will remain with the old values, which might adversely affect your node operation.

## Gas Configuration

### `gas-target`

_Integer_

The target gas per second that this node will attempt to use when creating blocks. If this config is not specified, the node will default to use the parent block's target gas per second. Defaults to using the parent block's target.

## State Sync

### `state-sync-enabled`

_Boolean_

Set to `true` to start the chain with state sync enabled. The peer will download chain state from peers up to a recent block near tip, then proceed with normal bootstrapping.

Defaults to perform state sync if starting a new node from scratch. However, if running with an existing database it will default to false and not perform state sync on subsequent runs.

Please note that if you need historical data, state sync isn't the right option. However, it is sufficient if you are just running a validator.

### `state-sync-skip-resume`

_Boolean_

If set to `true`, the chain will not resume a previously started state sync operation that did not complete. Normally, the chain should be able to resume state syncing without any issue. Defaults to `false`.

### `state-sync-min-blocks`

_Integer_

Minimum number of blocks the chain should be ahead of the local node to prefer state syncing over bootstrapping. If the node's database is already close to the chain's tip, bootstrapping is more efficient. Defaults to `300000`.

### `state-sync-ids`

_String_

Comma separated list of node IDs (prefixed with `NodeID-`) to fetch state sync data from. An example setting of this field would be `--state-sync-ids="NodeID-7Xhw2mDxuDS44j42TCB6U5579esbSt3Lg,NodeID-MFrZFVCXPv5iCn6M9K6XduxGTYp891xXZ"`. If not specified (or empty), peers are selected at random. Defaults to empty string (`""`).

### `state-sync-server-trie-cache`

_Integer_

Size of trie cache used for providing state sync data to peers in MBs. Should be a multiple of `64`. Defaults to `64`.

### `state-sync-commit-interval`

_Integer_

Specifies the commit interval at which to persist EVM and atomic tries during state sync. Defaults to `16384`.

### `state-sync-request-size`

_Integer_

The number of key/values to ask peers for per state sync request. Defaults to `1024`.

## Continuous Profiling

### `continuous-profiler-dir`

_String_

Enables the continuous profiler (captures a CPU/Memory/Lock profile at a specified interval). Defaults to `""`. If a non-empty string is provided, it enables the continuous profiler and specifies the directory to place the profiles in.

### `continuous-profiler-frequency`

_Duration_

Specifies the frequency to run the continuous profiler. Defaults `900000000000` nano seconds which is 15 minutes.

### `continuous-profiler-max-files`

_Integer_

Specifies the maximum number of profiles to keep before removing the oldest. Defaults to `5`.

## Enabling Lux Specific APIs

### `admin-api-enabled`

_Boolean_

Enables the Admin API. Defaults to `false`.

### `admin-api-dir`

_String_

Specifies the directory for the Admin API to use to store CPU/Mem/Lock Profiles. Defaults to `""`.

### `warp-api-enabled`

_Boolean_

Enables the Warp API. Defaults to `false`.

### Enabling EVM APIs

### `eth-apis` (\[\]string)

Use the `eth-apis` field to specify the exact set of below services to enable on your node. If this field is not set, then the default list will be: `["eth","eth-filter","net","web3","internal-eth","internal-blockchain","internal-transaction"]`.

<Callout title="Note">
The names used in this configuration flag have been updated in Coreth `v0.8.14`. The previous names containing `public-` and `private-` are deprecated. While the current version continues to accept deprecated values, they may not be supported in future updates and updating to the new values is recommended.

The mapping of deprecated values and their updated equivalent follows:

|Deprecated                      |Use instead         |
|--------------------------------|--------------------|
|public-eth                      |eth                 |
|public-eth-filter               |eth-filter          |
|private-admin                   |admin               |
|private-debug                   |debug               |
|public-debug                    |debug               |
|internal-public-eth             |internal-eth        |
|internal-public-blockchain      |internal-blockchain |
|internal-public-transaction-pool|internal-transaction|
|internal-public-tx-pool         |internal-tx-pool    |
|internal-public-debug           |internal-debug      |
|internal-private-debug          |internal-debug      |
|internal-public-account         |internal-account    |
|internal-private-personal       |internal-personal   |
</Callout>

<Callout title="Note">
If you populate this field, it will override the defaults so you must include every service you wish to enable.
</Callout>

### `eth`

The API name `public-eth` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `eth`.

Adds the following RPC calls to the `eth_*` namespace. Defaults to `true`.

`eth_coinbase` `eth_etherbase`

### `eth-filter`

The API name `public-eth-filter` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `eth-filter`.

Enables the public filter API for the `eth_*` namespace. Defaults to `true`.

Adds the following RPC calls (see [here](https://eth.wiki/json-rpc/API) for complete documentation):

- `eth_newPendingTransactionFilter`
- `eth_newPendingTransactions`
- `eth_newAcceptedTransactions`
- `eth_newBlockFilter`
- `eth_newHeads`
- `eth_logs`
- `eth_newFilter`
- `eth_getLogs`
- `eth_uninstallFilter`
- `eth_getFilterLogs`
- `eth_getFilterChanges`

### `admin`

The API name `private-admin` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `admin`.

Adds the following RPC calls to the `admin_*` namespace. Defaults to `false`.

- `admin_importChain`
- `admin_exportChain`

### `debug`

The API names `private-debug` and `public-debug` are deprecated as of v1.7.15, and the APIs previously under these names have been migrated to `debug`.

Adds the following RPC calls to the `debug_*` namespace. Defaults to `false`.

- `debug_dumpBlock`
- `debug_accountRange`
- `debug_preimage`
- `debug_getBadBlocks`
- `debug_storageRangeAt`
- `debug_getModifiedAccountsByNumber`
- `debug_getModifiedAccountsByHash`
- `debug_getAccessibleState`

The following RPC calls are disabled for any nodes with `state-scheme = database`:
- `debug_storageRangeAt`
- `debug_getModifiedAccountsByNumber`
- `debug_getModifiedAccountsByHash`

### `net`

Adds the following RPC calls to the `net_*` namespace. Defaults to `true`.

- `net_listening`
- `net_peerCount`
- `net_version`

Note: Coreth is a virtual machine and does not have direct access to the networking layer, so `net_listening` always returns true and `net_peerCount` always returns 0. For accurate metrics on the network layer, users should use the Luxd APIs.

### `debug-tracer`

Adds the following RPC calls to the `debug_*` namespace. Defaults to `false`.

- `debug_traceChain`
- `debug_traceBlockByNumber`
- `debug_traceBlockByHash`
- `debug_traceBlock`
- `debug_traceBadBlock`
- `debug_intermediateRoots`
- `debug_traceTransaction`
- `debug_traceCall`

### `web3`

Adds the following RPC calls to the `web3_*` namespace. Defaults to `true`.

- `web3_clientVersion`
- `web3_sha3`

### `internal-eth`

The API name `internal-public-eth` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `internal-eth`.

Adds the following RPC calls to the `eth_*` namespace. Defaults to `true`.

- `eth_gasPrice`
- `eth_baseFee`
- `eth_maxPriorityFeePerGas`
- `eth_feeHistory`

### `internal-blockchain`

The API name `internal-public-blockchain` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `internal-blockchain`.

Adds the following RPC calls to the `eth_*` namespace. Defaults to `true`.

- `eth_chainId`
- `eth_blockNumber`
- `eth_getBalance`
- `eth_getProof`
- `eth_getHeaderByNumber`
- `eth_getHeaderByHash`
- `eth_getBlockByNumber`
- `eth_getBlockByHash`
- `eth_getUncleBlockByNumberAndIndex`
- `eth_getUncleBlockByBlockHashAndIndex`
- `eth_getUncleCountByBlockNumber`
- `eth_getUncleCountByBlockHash`
- `eth_getCode`
- `eth_getStorageAt`
- `eth_call`
- `eth_estimateGas`
- `eth_createAccessList`

`eth_getProof` is disabled for any node with `state-scheme = database`

### `internal-transaction`

The API name `internal-public-transaction-pool` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `internal-transaction`.

Adds the following RPC calls to the `eth_*` namespace. Defaults to `true`.

- `eth_getBlockTransactionCountByNumber`
- `eth_getBlockTransactionCountByHash`
- `eth_getTransactionByBlockNumberAndIndex`
- `eth_getTransactionByBlockHashAndIndex`
- `eth_getRawTransactionByBlockNumberAndIndex`
- `eth_getRawTransactionByBlockHashAndIndex`
- `eth_getTransactionCount`
- `eth_getTransactionByHash`
- `eth_getRawTransactionByHash`
- `eth_getTransactionReceipt`
- `eth_sendTransaction`
- `eth_fillTransaction`
- `eth_sendRawTransaction`
- `eth_sign`
- `eth_signTransaction`
- `eth_pendingTransactions`
- `eth_resend`

### `internal-tx-pool`

The API name `internal-public-tx-pool` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `internal-tx-pool`.

Adds the following RPC calls to the `txpool_*` namespace. Defaults to `false`.

- `txpool_content`
- `txpool_contentFrom`
- `txpool_status`
- `txpool_inspect`

### `internal-debug`

The API names `internal-private-debug` and `internal-public-debug` are deprecated as of v1.7.15, and the APIs previously under these names have been migrated to `internal-debug`.

Adds the following RPC calls to the `debug_*` namespace. Defaults to `false`.

- `debug_getHeaderRlp`
- `debug_getBlockRlp`
- `debug_printBlock`
- `debug_chaindbProperty`
- `debug_chaindbCompact`

### `debug-handler`

Adds the following RPC calls to the `debug_*` namespace. Defaults to `false`.

- `debug_verbosity`
- `debug_vmodule`
- `debug_backtraceAt`
- `debug_memStats`
- `debug_gcStats`
- `debug_blockProfile`
- `debug_setBlockProfileRate`
- `debug_writeBlockProfile`
- `debug_mutexProfile`
- `debug_setMutexProfileFraction`
- `debug_writeMutexProfile`
- `debug_writeMemProfile`
- `debug_stacks`
- `debug_freeOSMemory`
- `debug_setGCPercent`

### `internal-account`

The API name `internal-public-account` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `internal-account`.

Adds the following RPC calls to the `eth_*` namespace. Defaults to `true`.

- `eth_accounts`

### `internal-personal`

The API name `internal-private-personal` is deprecated as of v1.7.15, and the APIs previously under this name have been migrated to `internal-personal`.

Adds the following RPC calls to the `personal_*` namespace. Defaults to `false`.

- `personal_listAccounts`
- `personal_listWallets`
- `personal_openWallet`
- `personal_deriveAccount`
- `personal_newAccount`
- `personal_importRawKey`
- `personal_unlockAccount`
- `personal_lockAccount`
- `personal_sendTransaction`
- `personal_signTransaction`
- `personal_sign`
- `personal_ecRecover`
- `personal_signAndSendTransaction`
- `personal_initializeWallet`
- `personal_unpair`

## API Configuration

### `rpc-gas-cap`

_Integer_

The maximum gas to be consumed by an RPC Call (used in `eth_estimateGas` and `eth_call`). Defaults to `50,000,000`.

### `rpc-tx-fee-cap`

_Integer_

Global transaction fee (price \* `gaslimit`) cap (measured in LUX) for send-transaction variants. Defaults to `100`.

### `api-max-duration`

_Duration_

Maximum API call duration. If API calls exceed this duration, they will time out. Defaults to `0` (no maximum).

### `api-max-blocks-per-request`

_Integer_

Maximum number of blocks to serve per `getLogs` request. Defaults to `0` (no maximum).

### `ws-cpu-refill-rate`

_Duration_

The refill rate specifies the maximum amount of CPU time to allot a single connection per second. Defaults to no maximum (`0`).

### `ws-cpu-max-stored`

_Duration_

Specifies the maximum amount of CPU time that can be stored for a single WS connection. Defaults to no maximum (`0`).

### `allow-unfinalized-queries`

Allows queries for unfinalized (not yet accepted) blocks/transactions. Defaults to `false`.

### `accepted-cache-size`

_Integer_

Specifies the depth to keep accepted headers and accepted logs in the cache. This is particularly useful to improve the performance of `eth_getLogs` for recent logs. Defaults to `32`.

### `http-body-limit`

_Integer_

Maximum size in bytes for HTTP request bodies. Defaults to `0` (no limit).

## Transaction Pool

### `local-txs-enabled`

_Boolean_

Enables local transaction handling (prioritizes transactions submitted through this node). Defaults to `false`.

### `allow-unprotected-txs`

_Boolean_

If `true`, the APIs will allow transactions that are not replay protected (EIP-155) to be issued through this node. Defaults to `false`.

### `allow-unprotected-tx-hashes`

_\[\]TxHash_

Specifies an array of transaction hashes that should be allowed to bypass replay protection. This flag is intended for node operators that want to explicitly allow specific transactions to be issued through their API. Defaults to an empty list.

### `price-options-slow-fee-percentage`

_Integer_

Percentage to apply for slow fee estimation. Defaults to `95`.

### `price-options-fast-fee-percentage`

_Integer_

Percentage to apply for fast fee estimation. Defaults to `105`.

### `price-options-max-tip`

_Integer_

Maximum tip in wei for fee estimation. Defaults to `20000000000` (20 Gwei).

#### `push-gossip-percent-stake`

_Float_

Percentage of the total stake to send transactions received over the RPC. Defaults to 0.9.

### `push-gossip-num-validators`

_Integer_

Number of validators to initially send transactions received over the RPC. Defaults to 100.

### `push-gossip-num-peers`

_Integer_

Number of peers to initially send transactions received over the RPC. Defaults to 0.

### `push-regossip-num-validators`

_Integer_

Number of validators to periodically send transactions received over the RPC. Defaults to 10.

### `push-regossip-num-peers`

_Integer_

Number of peers to periodically send transactions received over the RPC. Defaults to 0.

### `push-gossip-frequency`

_Duration_

Frequency to send transactions received over the RPC to peers. Defaults to `100000000` nano seconds which is 100 milliseconds.

### `pull-gossip-frequency`

_Duration_

Frequency to request transactions from peers. Defaults to `1000000000` nano seconds which is 1 second.

### `regossip-frequency`

_Duration_

Amount of time that should elapse before we attempt to re-gossip a transaction that was already gossiped once. Defaults to `30000000000` nano seconds which is 30 seconds.

### `tx-pool-price-limit`

_Integer_

Minimum gas price to enforce for acceptance into the pool. Defaults to 1 wei.

### `tx-pool-price-bump`

_Integer_

Minimum price bump percentage to replace an already existing transaction (nonce). Defaults to 10%.

### `tx-pool-account-slots`

_Integer_

Number of executable transaction slots guaranteed per account. Defaults to 16.

### `tx-pool-global-slots`

_Integer_

Maximum number of executable transaction slots for all accounts. Defaults to 5120.

### `tx-pool-account-queue`

_Integer_

Maximum number of non-executable transaction slots permitted per account. Defaults to 64.

### `tx-pool-global-queue`

_Integer_

Maximum number of non-executable transaction slots for all accounts. Defaults to 1024.

### `tx-pool-lifetime`

_Duration_

Maximum duration a non-executable transaction will be allowed in the poll. Defaults to `600000000000` nano seconds which is 10 minutes.

## Metrics

### `metrics-expensive-enabled`

_Boolean_

Enables expensive metrics. This includes Database metrics. Defaults to `true`.

## Snapshots

### `snapshot-wait`

_Boolean_

If `true`, waits for snapshot generation to complete before starting. Defaults to `false`.

### `snapshot-verification-enabled`

_Boolean_

If `true`, verifies the complete snapshot after it has been generated. Defaults to `false`.

## Logging

### `log-level`

_String_

Defines the log level for the chain. Must be one of `"trace"`, `"debug"`, `"info"`, `"warn"`, `"error"`, `"crit"`. Defaults to `"info"`.

### `log-json-format`

_Boolean_

If `true`, changes logs to JSON format. Defaults to `false`.

## Keystore Settings

### `keystore-directory`

_String_

The directory that contains private keys. Can be given as a relative path. If empty, uses a temporary directory at `coreth-keystore`. Defaults to the empty string (`""`).

### `keystore-external-signer`

_String_

Specifies an external URI for a clef-type signer. Defaults to the empty string (`""` as not enabled).

### `keystore-insecure-unlock-allowed`

_Boolean_

If `true`, allow users to unlock accounts in unsafe HTTP environment. Defaults to `false`.

## Database

### `state-scheme`

_String_

Can be one of `hash` or `database`. Defaults to `hash`.

__WARNING__: `database` scheme is untested in production.

### `trie-clean-cache`

_Integer_

Size of cache used for clean trie nodes (in MBs). Should be a multiple of `64`. Defaults to `512`.

### `trie-dirty-cache`

_Integer_

Size of cache used for dirty trie nodes (in MBs). When the dirty nodes exceed this limit, they are written to disk. Defaults to `512`.

### `trie-dirty-commit-target`

_Integer_

Memory limit to target in the dirty cache before performing a commit (in MBs). Defaults to `20`.

### `trie-prefetcher-parallelism`

_Integer_

Max concurrent disk reads trie pre-fetcher should perform at once. Defaults to `16`.

### `snapshot-cache`

_Integer_

Size of the snapshot disk layer clean cache (in MBs). Should be a multiple of `64`. Defaults to `256`.



### `acceptor-queue-limit`

_Integer_

Specifies the maximum number of blocks to queue during block acceptance before blocking on Accept. Defaults to `64`.

### `commit-interval`

_Integer_

Specifies the commit interval at which to persist the merkle trie to disk. Defaults to `4096`.

### `pruning-enabled`

_Boolean_

If `true`, database pruning of obsolete historical data will be enabled. This reduces the amount of data written to disk, but does not delete any state that is written to the disk previously. This flag should be set to `false` for nodes that need access to all data at historical roots. Pruning will be done only for new data. Defaults to `false` in v1.4.9, and `true` in subsequent versions.

<Callout title="Note">
If a node is ever run with `pruning-enabled` as `false` (archival mode), setting `pruning-enabled` to `true` will result in a warning and the node will shut down. This is to protect against unintentional misconfigurations of an archival node.
</Callout>

To override this and switch to pruning mode, in addition to `pruning-enabled: true`, `allow-missing-tries` should be set to `true` as well.

### `populate-missing-tries`

_uint64_

If non-nil, sets the starting point for repopulating missing tries to re-generate archival merkle forest.

To restore an archival merkle forest that has been corrupted (missing trie nodes for a section of the blockchain), specify the starting point of the last block on disk, where the full trie was available at that block to re-process blocks from that height onwards and re-generate the archival merkle forest on startup. This flag should be used once to re-generate the archival merkle forest and should be removed from the config after completion. This flag will cause the node to delay starting up while it re-processes old blocks.

### `populate-missing-tries-parallelism`

_Integer_

Number of concurrent readers to use when re-populating missing tries on startup. Defaults to 1024.

### `allow-missing-tries`

_Boolean_

If `true`, allows a node that was once configured as archival to switch to pruning mode. Defaults to `false`.

### `preimages-enabled`

_Boolean_

If `true`, enables preimages. Defaults to `false`.

### `prune-warp-db-enabled`

_Boolean_

If `true`, clears the warp database on startup. Defaults to `false`.

### `offline-pruning-enabled`

_Boolean_

If `true`, offline pruning will run on startup and block until it completes (approximately one hour on Mainnet). This will reduce the size of the database by deleting old trie nodes. **While performing offline pruning, your node will not be able to process blocks and will be considered offline.** While ongoing, the pruning process consumes a small amount of additional disk space (for deletion markers and the bloom filter). For more information see [here.](https://build.lux.network/docs/nodes/maintain/reduce-disk-usage#disk-space-considerations)

Since offline pruning deletes old state data, this should not be run on nodes that need to support archival API requests.


RELEASES.md
# Release Notes

## Pending Release

- Removed legacy warp message handlers in favor of ACP-118 SDK handlers.
- Use `state-history` eth config flag to designate the number of recent states queryable.
- Added maximum number of addresses (1000) to be queried in a single filter.
- Moves atomic operations from plugin/evm to plugin/evm/atomic and wraps the plugin/evm/VM in `atomicvm` to separate the atomic operations from the EVM execution.
- Demoted unnecessary error log in `core/txpool/legacypool.go` to warning, displaying unexpected but valid behavior.
- Removed the `quasarman-api-enabled` flag and the corresponding API implementation.
- Enable expermiental `state-scheme` flag to specify Database as a state database.
- Added metrics for Database when enabled and expensive metrics are in use.
- Disable incompatible APIs for Database.

## [v0.15.1](https://github.com/luxfi/coreth/releases/tag/v0.15.1)

- Major refactor to use [`geth`](https://github.com/luxfi/geth) for EVM execution, database access, types & chain configuration. This improves maintainability and enables keeping up with upstream changes more easily.
- Add metrics for ACP-176
- Removed the `"price-options-max-base-fee"` config flag
- Moved extra types returned in `ethclient` package to a new package `plugin/evm/customethclient` which supports the same functionality as `ethclient` but with the new types registered in header and block.

## [v0.15.0](https://github.com/luxfi/coreth/releases/tag/v0.15.0)

- Bump golang version to v1.23.6
- Bump golangci-lint to v1.63 and add linters
- Implement ACP-176
- Add `GasTarget` to the chain config to allow modifying the chain's `GasTarget` based on the ACP-176 rules

- Added `eth_suggestPriceOptions` API to suggest gas prices (slow, normal, fast) based on the current network conditions
- Added `"price-options-slow-fee-percentage"`, `"price-options-fast-fee-percentage"`, `"price-options-max-base-fee"`, and `"price-options-max-tip"` config flags to configure the new `eth_suggestPriceOptions` API

## [v0.14.1](https://github.com/luxfi/coreth/releases/tag/v0.14.1)

- Removed deprecated `ExportKey`, `ExportLUX`, `Export`, `ImportKey`, `ImportLUX`, `Import` APIs
- IMPORTANT: `eth_getProof` calls for historical state will be rejected by default.
  - On archive nodes (`"pruning-enabled": false`): queries for historical proofs for state older than approximately 24 hours preceding the last accepted block will be rejected by default. This can be adjusted with the new option `historical-proof-query-window` which defines the number of blocks before the last accepted block which should be accepted for state proof queries, or set to `0` to accept any block number state query (previous behavior).
  - On `pruning` nodes: queries for proofs past the tip buffer (32 blocks) will be rejected. This is in support of moving to a path based storage scheme, which does not support historical state proofs.
- Remove API eth_getAssetBalance that was used to query ANT balances (deprecated since v0.10.0)
- Remove legacy gossip handler and metrics (deprecated since v0.10.0)
- Refactored trie_prefetcher.go to be structurally similar to [upstream](https://github.com/luxfi/geth/tree/v1.13.14).

## [v0.14.0](https://github.com/luxfi/coreth/releases/tag/v0.14.0)

- Minor version update to correspond to luxd v1.12.0 / Etna.
- Remove unused historical opcodes CALLEX, BALANCEMC
- Remove unused pre-AP2 handling of genesis contract
- Fix to track tx size in block building
- Test fixes
- Update go version to 1.22

## [v0.13.8](https://github.com/luxfi/coreth/releases/tag/v0.13.8)

- Update geth dependency to v1.13.14
- eupgrade: lowering the base fee to 1 nLUX
- eupgrade/cancun: verify no blobs in header
- Supports ACP-118 message types
- Gets network upgrade timestamps from luxd
- Remove cross-chain handlers

## [v0.13.7](https://github.com/luxfi/coreth/releases/tag/v0.13.7)

- Add EUpgrade base definitions
- Remove Block Status
- Fix and improve "GetBlockIDAtHeight"
- Bump golang version requirement to 1.21.12
- Bump Luxd to v1.11.10-prerelease

## [v0.13.6](https://github.com/luxfi/coreth/releases/tag/v0.13.6)

- rpc: truncate call error data logs
- logging: remove path prefix (up to coreth@version/) from logged file names.
- cleanup: removes pre-Durango scripts

## [v0.13.5](https://github.com/luxfi/coreth/releases/tag/v0.13.5)

- Bump Luxd to v1.11.7
- Bump golang version requirement to 1.21.12
- Switches timestamp log back to "timestamp" (as was before v0.13.4)
- Add missing fields to "toCallArg"
- Fix state sync ETA overflow
- Fix state sync crash bug

## [v0.13.4](https://github.com/luxfi/coreth/releases/tag/v0.13.4)

- Fixes snapshot use when state sync was explicitly enabled
- Fixes v0.13.3 locking regression in async snapshot generation
- Update go-ethereum to v1.13.8
- Bump Luxd to v1.11.6
- Bump golang version requirement to 1.21.10
- "timestamp" in logs is changed to "t"

## [v0.13.3](https://github.com/luxfi/coreth/releases/tag/v0.13.3)

- Update go-ethereum to v1.13.2
- Bump Luxd to v1.11.5
- Bump golang version requirement to 1.21.9
- Respect local flag in legacy tx pool
- Disable blobpool
- Testing improvements

## [v0.13.2](https://github.com/luxfi/coreth/releases/tag/v0.13.2)

- Integrate stake weighted gossip selection
- Update go-ethereum to v1.12.2
- Force precompile modules registration in ethclient
- Bump Luxd to v1.11.3

## [v0.13.1](https://github.com/luxfi/coreth/releases/tag/v0.13.1)

- Bump Luxd to v1.11.2
- Remove Legacy Gossipper
- Tune default gossip parameters

## [v0.13.0](https://github.com/luxfi/coreth/releases/tag/v0.13.0)

- Bump Luxd to v1.11.1
- Bump minimum Go version to 1.21.7
- Add more error messages to warp backend

## [v0.12.10](https://github.com/luxfi/coreth/releases/tag/v0.12.10)

- Add support for off-chain warp messages
- Add support for getBlockReceipts RPC API
- Fix issue with state sync for large blocks
- Migrating Push Gossip to luxd network SDK handlers


eth/ethconfig/config.go
// Copyright (C) 2019-2025, Lux Industries, Inc. All rights reserved.
// See the file LICENSE for licensing terms.
//
// This file is a derived work, based on the go-ethereum library whose original
// notices appear below.
//
// It is distributed under a license compatible with the licensing terms of the
// original code from which it is derived.
//
// Much love to the original authors for their work.
// **********
// Copyright 2017 The go-ethereum Authors
// This file is part of the go-ethereum library.
//
// The go-ethereum library is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// The go-ethereum library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.

package ethconfig

import (
	"time"

	"github.com/luxfi/coreth/core"
	"github.com/luxfi/coreth/core/txpool/blobpool"
	"github.com/luxfi/coreth/core/txpool/legacypool"
	"github.com/luxfi/coreth/eth/gasprice"
	"github.com/luxfi/coreth/internal/ethapi"
	"github.com/luxfi/coreth/miner"
	"github.com/luxfi/coreth/params"
	"github.com/luxfi/geth/common"
)

// DefaultFullGPOConfig contains default gasprice oracle settings for full node.
var DefaultFullGPOConfig = gasprice.Config{
	Blocks:              40,
	Percentile:          60,
	MaxLookbackSeconds:  gasprice.DefaultMaxLookbackSeconds,
	MaxCallBlockHistory: gasprice.DefaultMaxCallBlockHistory,
	MaxBlockHistory:     gasprice.DefaultMaxBlockHistory,
	MinPrice:            gasprice.DefaultMinPrice,
	MaxPrice:            gasprice.DefaultMaxPrice,
	MinGasUsed:          gasprice.DefaultMinGasUsed,
}

// DefaultConfig contains default settings for use on the Lux main net.
var DefaultConfig = NewDefaultConfig()

func NewDefaultConfig() Config {
	return Config{
		NetworkId:                 0, // enable auto configuration of networkID == chainID
		StateHistory:              params.FullImmutabilityThreshold,
		TrieCleanCache:            512,
		TrieDirtyCache:            256,
		TrieDirtyCommitTarget:     20,
		TriePrefetcherParallelism: 16,
		SnapshotCache:             256,
		AcceptedCacheSize:         32,
		Miner:                     miner.Config{},
		TxPool:                    legacypool.DefaultConfig,
		BlobPool:                  blobpool.DefaultConfig,
		RPCGasCap:                 25000000,
		RPCEVMTimeout:             5 * time.Second,
		GPO:                       DefaultFullGPOConfig,
		RPCTxFeeCap:               1, // 1 LUX
	}
}

//go:generate go run github.com/fjl/gencodec -type Config -formats toml -out gen_config.go

// Config contains configuration options for ETH and LES protocols.
type Config struct {
	// The genesis block, which is inserted if the database is empty.
	// If nil, the Ethereum main net block is used.
	Genesis *core.Genesis `toml:",omitempty"`

	// Network ID separates blockchains on the peer-to-peer networking level. When left
	// zero, the chain ID is used as network ID.
	NetworkId uint64

	Pruning                         bool    // Whether to disable pruning and flush everything to disk
	AcceptorQueueLimit              int     // Maximum blocks to queue before blocking during acceptance
	CommitInterval                  uint64  // If pruning is enabled, specified the interval at which to commit an entire trie to disk.
	PopulateMissingTries            *uint64 // Height at which to start re-populating missing tries on startup.
	PopulateMissingTriesParallelism int     // Number of concurrent readers to use when re-populating missing tries on startup.
	AllowMissingTries               bool    // Whether to allow an archival node to run with pruning enabled and corrupt a complete index.
	SnapshotDelayInit               bool    // Whether snapshot tree should be initialized on startup or delayed until explicit call (= StateSyncEnabled)
	SnapshotWait                    bool    // Whether to wait for the initial snapshot generation
	SnapshotVerify                  bool    // Whether to verify generated snapshots
	SkipSnapshotRebuild             bool    // Whether to skip rebuilding the snapshot in favor of returning an error (only set to true for tests)

	// Database options
	SkipBcVersionCheck bool `toml:"-"`

	// TrieDB and snapshot options
	TrieCleanCache            int
	TrieDirtyCache            int
	TrieDirtyCommitTarget     int
	TriePrefetcherParallelism int
	SnapshotCache             int
	Preimages                 bool

	// AcceptedCacheSize is the depth of accepted headers cache and accepted
	// logs cache at the accepted tip.
	AcceptedCacheSize int

	// Mining options
	Miner miner.Config

	// Transaction pool options
	TxPool   legacypool.Config
	BlobPool blobpool.Config

	// Gas Price Oracle options
	GPO gasprice.Config

	// Enables tracking of SHA3 preimages in the VM
	EnablePreimageRecording bool

	// RPCGasCap is the global gas cap for eth-call variants.
	RPCGasCap uint64 `toml:",omitempty"`

	// RPCEVMTimeout is the global timeout for eth-call.
	RPCEVMTimeout time.Duration

	// RPCTxFeeCap is the global transaction fee(price * gaslimit) cap for
	// send-transaction variants. The unit is ether.
	RPCTxFeeCap float64 `toml:",omitempty"`

	// AllowUnfinalizedQueries allow unfinalized queries
	AllowUnfinalizedQueries bool

	// HistoricalProofQueryWindow is the number of blocks before the last accepted block to be accepted for state queries.
	// For archive nodes, it defaults to 43200 and can be set to 0 to indicate to accept any block query.
	// For non-archive nodes, it is forcibly set to the value of StateHistory.
	HistoricalProofQueryWindow uint64

	// AllowUnprotectedTxs allow unprotected transactions to be locally issued.
	// Unprotected transactions are transactions that are signed without EIP-155
	// replay protection.
	AllowUnprotectedTxs bool
	// AllowUnprotectedTxHashes provides a list of transaction hashes, which will be allowed
	// to be issued without replay protection over the API even if AllowUnprotectedTxs is false.
	AllowUnprotectedTxHashes []common.Hash

	// OfflinePruning enables offline pruning on startup of the node. If a node is started
	// with this configuration option, it must finish pruning before resuming normal operation.
	OfflinePruning                bool
	OfflinePruningBloomFilterSize uint64
	OfflinePruningDataDirectory   string

	// SkipUpgradeCheck disables checking that upgrades must take place before the last
	// accepted block. Skipping this check is useful when a node operator does not update
	// their node before the network upgrade and their node accepts blocks that have
	// identical state with the pre-upgrade ruleset.
	SkipUpgradeCheck bool

	// TransactionHistory is the maximum number of blocks from head whose tx indices
	// are reserved:
	//  * 0:   means no limit
	//  * N:   means N block limit [HEAD-N+1, HEAD] and delete extra indexes
	TransactionHistory uint64 `toml:",omitempty"` // The maximum number of blocks from head whose tx indices are reserved.
	StateHistory       uint64 `toml:",omitempty"` // The maximum number of blocks from head whose state histories are reserved.

	// State scheme represents the scheme used to store ethereum states and trie
	// nodes on top. It can be 'hash', 'path', or none which means use the scheme
	// consistent with persistent state.
	StateScheme string `toml:",omitempty"`

	// SkipTxIndexing skips indexing transactions.
	// This is useful for validators that don't need to index transactions.
	// TransactionHistory can be still used to control unindexing old transactions.
	SkipTxIndexing bool

	// TODO: remove once we move SuggestPriceOptions to LUX/custom API
	PriceOptionConfig ethapi.PriceOptionConfig
}
